lib/github_advisory_sync.rb

```diff
@@ -70,88 +59,102 @@ def github_api(adapter = :net_http)
-      end
+
+      raise(GitHubGraphQLAPIError, body_obj["errors"].map { |e| e["message"] }.join(", ")) if body_obj["errors"]
+
       body_obj
     end
 
-    def retrieve_all_github_advisories(max_pages = 1000, page_size = 100)
-      all_advisories = []
+    def all_rubygem_advisories
+      advisories = {}
+
+      retrieve_all_rubygem_vulnerabilities.each do |vulnerability|
+        advisory = GitHubAdvisory.new(vulnerability["advisory"])
+
+        next if advisory.withdrawn?
+
+        advisories[advisory.primary_id] = advisory unless advisories.key? advisory.primary_id
+
+        advisories[advisory.primary_id].vulnerabilities << vulnerability.except("advisory")
+      end
+
+      advisories.values
+    end
+
+    def retrieve_all_rubygem_vulnerabilities(max_pages = 1000, page_size = 100)
+      all_vulnerabilities = []
       variables = { "first" => page_size }
       max_pages.times do |page_num|
-        puts "Getting page #{page_num + 1} of GitHub Advisories"
-        page = github_graphql_query(:GITHUB_ADVISORIES_WITH_RUBYGEM_VULNERABILITY, variables)
-        advisories_this_page = page["data"]["securityAdvisories"]["nodes"]
-        all_advisories += advisories_this_page
-        break unless page["data"]["securityAdvisories"]["pageInfo"]["hasNextPage"] == true
-        variables["after"] = page["data"]["securityAdvisories"]["pageInfo"]["endCursor"]
-      end
-      puts "Retrieved #{all_advisories.length} Advisories from GitHub API"
+        puts "Getting page #{page_num + 1} of GitHub Vulnerabilities"
+
+        page = github_graphql_query(:RUBYGEM_VULNERABILITIES_WITH_GITHUB_ADVISORIES, variables)
+        vulnerabilities_this_page = page["data"]["securityVulnerabilities"]["nodes"]
+        all_vulnerabilities += vulnerabilities_this_page
 
-      all_advisories.map do |advisory_graphql_obj|
-        GitHubAdvisory.new github_advisory_graphql_object: advisory_graphql_obj
+        break unless page["data"]["securityVulnerabilities"]["pageInfo"]["hasNextPage"] == true
+
+        variables["after"] = page["data"]["securityVulnerabilities"]["pageInfo"]["endCursor"]
       end
-    end
+      puts "Retrieved #{all_vulnerabilities.length} Vulnerabilities from GitHub API"
 
-    def retrieve_all_rubygem_publishable_advisories
-      all_advisories = retrieve_all_github_advisories
-      # remove withdrawn advisories,
-      # and remove those where there are no vulnerabilities for ruby
-      all_advisories.reject { |advisory| advisory.withdrawn? }
-                    .select { |advisory| advisory.has_ruby_vulnerabilities? }
+      all_vulnerabilities
     end
 
     module GraphQLQueries
-      GITHUB_ADVISORIES_WITH_RUBYGEM_VULNERABILITY = <<-GRAPHQL.freeze
+      RUBYGEM_VULNERABILITIES_WITH_GITHUB_ADVISORIES = <<-GRAPHQL.freeze
         query($first: Int, $after: String) {
-          securityAdvisories(first: $first, after: $after) {
+          securityVulnerabilities(first: $first, after: $after, ecosystem:RUBYGEMS) {
```

Comment:
Very good! We'd recently moved ecosystem down to the `vulnerability` node, and I'm glad you noticed this here.

Seeing this tho, I wonder two things @SkipKayhil:

- life would be easier if offered an `ecosystem` filter on securityAdvisories, eh?
- would it be useful to provide an `updated_after` filter, so we don't have to page thru every single advisory? that way maybe we could just store a timestamp of the last time the importer was run.

Reply:
Both of those sounds like great additions! I think adding ecosystem to `securityAdvisories` would be a huge win for sure, and probably remove a lot of the refactoring I did to invert the relationship

Reply:
We're probably going to touch the API this summer, so I'll add this to my list of tweaks. If you have any other suggestions, holler.